src/main/host/descriptor/timerfd.rs

```diff
@@ -0,0 +1,275 @@
+use crate::host::syscall::io::{IoVec, IoVecWriter};
+use crate::host::syscall_types::{SyscallError, SyscallResult};
+use crate::host::timer::Timer;
+use crate::utility::callback_queue::{CallbackQueue, Handle};
+use crate::utility::HostTreePointer;
+
+pub struct TimerFd {
+    timer: Timer,
+    event_source: StateEventSource,
+    status: FileStatus,
+    state: FileState,
+    // Should only be used by `OpenFile` to make sure there is only ever one `OpenFile` instance for
+    // this file,
+    has_open_file: bool,
+}
+
+impl TimerFd {
+    /// Creates a new [`TimerFd`] object that internally sets up a [`Timer`] that can be waited on
+    /// with poll, select, and epoll, enabling support for timerfd_create(2).
+    ///
+    /// We wrap the new [`TimerFd`] in an [`Arc<AtomicRefCell>`] because we need to use a weak
+    /// reference to internally support setting up callback functions that reference the [`TimerFd`]
+    /// on timer expiration.
+    pub fn new(status: FileStatus) -> Arc<AtomicRefCell<Self>> {
+        // We need a circular reference here, so that the inner Timer can refer back to the outer
+        // TimerFd when executing a callback that will mutate the TimerFd when the timer expires.
+        Arc::new_cyclic(|weak| {
+            let weak_cloned = weak.clone();
+            AtomicRefCell::new(Self {
+                timer: Timer::new(move |_host| Self::timer_expired(&weak_cloned)),
+                event_source: StateEventSource::new(),
+                state: FileState::ACTIVE,
+                status,
+                has_open_file: false,
+            })
+        })
+    }
+
+    /// Called by the inner [`Timer`] when a timer expiration occurs.
+    fn timer_expired(timerfd_weak: &Weak<AtomicRefCell<TimerFd>>) {
+        let Some(timerfd) = timerfd_weak.upgrade() else {
+            log::trace!("Expired TimerFd no longer exists.");
+            return;
+        };
+
+        // The TimerFd may have become readable now that a timer expired. We use the CallbackQueue
+        // here to make sure that any listeners that need to wake up and handle a readable TimerFd
+        // are not invoked until after we release the borrow.
+        CallbackQueue::queue_and_run(|cb_queue| {
+            timerfd.borrow_mut().update_state(cb_queue);
+        });
+    }
+
+    /// Returns the number of expirations that have occured since the timer was last armed.
+    fn get_timer_count(&self) -> u64 {
+        self.timer.expiration_count()
+    }
+
+    /// Returns the relative duration until the next expiration event occurs if the timer is armed,
+    /// and `None` if the timer is disarmed.
+    pub fn get_timer_remaining(&self) -> Option<SimulationTime> {
+        self.timer.remaining_time()
+    }
+
+    /// Returns the relative duration over which the timer has been configured to periodically
+    /// expire, which may be zero if the timer is only configured to expire once.
+    pub fn get_timer_interval(&self) -> SimulationTime {
```

Comment:
Maybe change this and `expire_interval` to return an `Option<SimulationTime>`. Or drop a TODO comment to do so and limit this PR to a more straightforward port of the existing C code.